### DISCUSSION TITLE
Resource Content / LLM Context Placement Hint

### DISCUSSION THREAD

**Discussion by evalstate** [modelcontextprotocol/modelcontextprotocol#90] [Ideas - General]
### Pre-submission Checklist

- [X] I have verified this would not be more appropriate as a feature request in a specific repository
- [X] I have searched existing discussions to avoid duplicates

### Your Idea

~~While MCP includes a Priority/Audience annotation, it would be nice to be able to have an explicit "Hint" to indicate whether a Resource is intended to be placed within LLM context or as an attachment.~~  

Edit: The Audience annotation on Resources does this - instead I think some guidance on this topic would be appropriate. 

I do think it would make sense to allow the return of a "Resource Pointer" Server the option of returning a Resource without necessarily embedding in the response. Of course, the Server could simply list Resources in it's textual response, but I think this would be cleaner.

We should also distinguish between the intended difference of returning an ImageContent as part of the CallToolResult vs returning it an EmbeddedResource. 

A common case for this distinction would be:

1. Image Generation. User uses a tool to create an image from a prompt for viewing. In this case putting it in the LLM Context Window is unnecessary and consumes space. This is how HuggingFace chat-ui and ChatGPT behave.
1. Image Analysis. User references an image - e.g. has a tool retrieve it from a URL and requests analysis/OCR. In this case it is intended to be included in the Context Window for the next step. This is the current default behaviour for Claude Desktop.

Although ultimately a Client handling decision, I think in a lot of cases the Client and/or Server would be able to provide a hint for the expected handling of a Resource. 

While the protocol provides mechanisms like Roots and Resource Templates for organizing and accessing resources, it doesn't currently offer a straightforward way to communicate these handling preferences between Client and Server.

Some ideas:
 - Extend ResourceContents to have an optional contextPlacement indicator ("prefer-context")
 - Include ResourceMetadata to have a preferred placement ("context" or "attachment").

**Example Flows:**

```mermaid
sequenceDiagram
    participant U as User
    participant C as Client
    participant L as LLM
    participant S as Server

    rect rgb(240, 240, 240)
        Note over U,S: Claude Desktop: Current Flow - All Images In Context
        U->>C: "Generate beach image"
        C->>L: Send prompt
        L->>C: Request tool 'generate_image'
        C->>S: tools/call Request
        S-->>C: Response with ImageContent
        C->>L: Include image in context
        C-->>U: Display image
    end

    rect rgb(230, 240, 250)
        Note over U,S: Desired Flow - Generated Images
        U->>C: "Generate beach image"
        C->>L: Send prompt
        L->>C: Request tool 'generate_image'
        C->>S: tools/call Request
        S-->>C: Response with ImageContent<br/>(hint: prefer-attachment)
        C-->>U: Display image
        Note over C: Image not added<br/>to LLM context
        C->>L: Continue conversation
    end

    rect rgb(230, 250, 240)
        Note over U,S: Desired Flow - Analysis Case (e.g. OCR)
        U->>C: "What's in this image?"
        C->>L: Send prompt
        L->>C: Request tool 'analyze_image'
        C->>S: tools/call Request
        S-->>C: Response with ImageContent<br/>(hint: prefer-context)
        C->>L: Include image in context
        L-->>C: Image analysis
        C-->>U: Display analysis
    end
 ```

In this example, the create_image tool has been hardcoded to generate beach scenes. Claude Desktop recognises the image is not as requested (and even retries!). This shows the Client incorporating the resource in to the Context Window.

![image](https://github.com/user-attachments/assets/99915525-52ef-4488-acb8-4f0690e63ab7)

### Scope

- [ ] Protocol Specification
- [ ] SDK Features
- [ ] Documentation
- [ ] Developer Experience
- [ ] Other

**Comment by jonathanhefner** [comment]
I agree it would be good to get some guidance in the spec.  There is also:

* `AudioContent` vs `EmbeddedResource` with an audio MIME type
* `TextContent` vs `EmbeddedResource` with a `text/plain` MIME type
* Potentially `TextContent` with a MIME type vs `EmbeddedResource` with a MIME type (see https://github.com/modelcontextprotocol/modelcontextprotocol/discussions/180#discussioncomment-12614185)
* Potentially `DataContent` vs `EmbeddedResource` with an `application/json` MIME type (see https://github.com/modelcontextprotocol/modelcontextprotocol/pull/356)

Plus, as mentioned, the `audience` annotation, which is applicable to all of the above.

> Some ideas:
>
> * Extend ResourceContents to have an optional contextPlacement indicator ("prefer-context")
> * Include ResourceMetadata to have a preferred placement ("context" or "attachment").

Could you explain a bit more about how that would differ from the `audience` annotation?

Also, if `TextContent` supported a MIME type and a `DataContent` type were introduced, is it possible that the `EmbeddedResource` type in and of itself could denote preferred placement?

**Reply by evalstate** [reply]
My view is that ->

TextContent | ImageContent | AudioContent are part of the User/Assistant conversation and destined to be tokenized by the LLM API (hence why I am keen that `TextContent` remains as `text/plain` in all circumstances). 

EmbeddedResource gives you a way to "attach" arbitrary content types - which could include PDFs, JSON, XML etc. My assumption here is that it denotes an expectation of "handling" by the Host/Client application before presentation to the LLM. So for example, attaching a PDF as a Document to the Completions API, or filtering out some JSON content to a known schema.

There's a level of ambiguity here, but I think that rough mental model should fit most conversations. But for the purpose of explaining the overlap - ImageContent is part of the conversation and ends up with the LLM. EmbeddedResource containing an Image with an audience of `user` would be rendered to the User, but with no hard expectation of it ending up in the LLMs context. This gives a flexible way to manage a range of Content types and indicate certain content for User Display/Host Processing but not necessarily LLM context.

Since writing that comment on Nov 30 last year you may have missed the edit I made at the top where I struck through the initial assertion and instead said:

> Edit: The Audience annotation on Resources does this - instead I think some guidance on this topic would be appropriate.

So if the Annotation is biased towards the User we are saying it is most important to render to the Human, or have the Host/Client application deal with it.  If biased towards the Assistant we are saying it is most important to ensure it is tokenized in some way for the LLM.

This https://github.com/modelcontextprotocol/modelcontextprotocol/pull/223 PR aims to close some of that loop, with suggested wording to clarify:

> For Resources annotated with `audience=user, priority=1` the Host MAY choose not to present the full content of the Resource to the LLM.

**Reply by evalstate** [reply]
See also https://github.com/modelcontextprotocol/modelcontextprotocol/pull/415 which aims to bring the guidance more in line with RFC3986 (referenced on the Resources page at the top)

**Reply by jonathanhefner** [reply]
> There's a level of ambiguity here, but I think that rough mental model should fit most conversations. But for the purpose of explaining the overlap - ImageContent is part of the conversation and ends up with the LLM. EmbeddedResource containing an Image with an audience of `user` would be rendered to the User, but with no hard expectation of it ending up in the LLMs context. This gives a flexible way to manage a range of Content types and indicate certain content for User Display/Host Processing but not necessarily LLM context.

To me, it seems like `audience` is the key distinguisher.  `EmbeddedResource` vs `ImageContent`, `AudioContent`, etc doesn't seem to be a meaningful distinction.

Do you think things would be more clear if we had a unified content item type with appropriate metadata?  Something like:

  ```ts
  export interface ContentItem {
    mimeType: string

    // Plain text or base64-encoded binary, depending on `mimeType`.
    data: string

    // All annotations are optional.
    annotations?: {
      // "base64" if `data` is base64-encoded binary; otherwise null.
      // If omitted, client may assume `data` is plain text.
      encoding?: "base64" | null;

      // Indicates whether the item is for user or model or both.
      // If for model, the item should be added to the context.
      // If omitted, client may assume the item is for both user and model.
      audience?: Role[];

      // Hint for the client about how the item should be rendered.
      // If omitted, client may assume "inline".
      //
      // Possibly support other values here, e.g. "ui" for dynamic UI widgets?
      disposition?: "inline" | "attachment";

      // Description of the item, targeted at either the user or the model (or both).
      //
      // Possibly require when `disposition: "attachment"` for use as a label?
      description?: string;

      // URL for the item.  May be an ephemeral URL.
      url?: string;

      // Hint for the client that `url` is ephemeral.
      // Only a hint -- false does not guarantee that the item will be accessible
      // via the URL in the future, and true does not guarantee that the item
      // will become inaccessible.
      ephemeral?: boolean;

      // A reference to a schema that `data` conforms to.
      // Neither the format of the reference nor the format of the schema is
      // specified.  If the receiver recognizes the format of the reference and
      // the format of the schema, then the schema may be used for validation;
      // however, the primary purpose of `schemaRef` is to identify the shape of
      // `data` beyond what `mimeType` is able to convey.
      //
      // Example: a canonical URL of a JSON schema when `mimeType: "application/json"`.
      //
      // Possibly provide more guidance here for popular MIME types?
      schemaRef?: string;
    }
  }
  ```

The above `ContentItem` type would replace `EmbeddedResource` and all `*Content` types.

**Reply by evalstate** [reply]
Personal opinion - yes, something like this would be an improvement, and also mean that future modalities such as Video would slot-in without any real change. This scheme would also be supportable for a transitional period where API compatibility could be maintained. 

There's plenty to discuss over the specifics of what that may look like, as well as some stronger guidance over the transmission of "files" to be done at the same time I would say.  

More personal opinion - this isn't something that needs to be rushed: the status quo isn't restrictive - the missing thing is understanding of the existing options. As HTTP Streaming becomes more established for example, it would be useful to have some time to consider what `roots` mean in that context and whether the `file://` uri restriction should be removed or have MCP Servers be able to request "Root Resources" from the MCP Client. This is the kind of thing where I think a Community Working Group could add a lot of value as real-world experience is gained, and we can consider the end-to-end effects of proposals from Spec -> SDK -> Hosting -> Client -> Host -> Human.  This is also an area where a couple of light touches like #223 can help.

The one thing I _would_ argue strongly to maintain is a simple default TextContent type that is known to be `text/plain` and used as the general default.

**Reply by evalstate** [reply]
Keeping it with this thread (I'll need to search again later), there was also an important conversation about just using the uri as a pointer. So, for example, rather than have the text|blob in the EmbeddedResource you would instead just send the uri with the expectation the Client would fetch it. This saves transmitting potentially large amounts of data as part of a CallToolResult (so for example you might generate an image, and instead of using EmbeddedResource you might return AttachedResource with a uri of `https://image-server/whatever.png`. I'll see if I can find the original thread.

**Reply by jonathanhefner** [reply]
> Keeping it with this thread (I'll need to search again later), there was also an important conversation about just using the uri as a pointer. So, for example, rather than have the text|blob in the EmbeddedResource you would instead just send the uri with the expectation the Client would fetch it. This saves transmitting potentially large amounts of data as part of a CallToolResult (so for example you might generate an image, and instead of using EmbeddedResource you might return AttachedResource with a uri of `https://image-server/whatever.png`. I'll see if I can find the original thread.

Ah, that's an interesting use case!  That could be very useful in agentic workflows where there may be several layers of message forwarding.

Then I think it would make sense to move `url` to the top-level alongside `data`.  (And add a stipulation that at least one of the two must be present, and if both are present, they should represent the same content.)

---

Adding some notes here:

* Instead of the `ephemeral` hint, support a `urlScope` hint with possible values like:
  * `"local"` - URL is for a local resource, e.g. a file (no liveness guarantee).
  * `"remote"` - URL is for a remote resource, i.e. requires network access (no liveness guarantee).
  * `"server"` - URL is for a server resource, i.e. can be fetched with `resources/read` (no liveness guarantee).
    * Possibly add a `"server/listed"` value that does guarantee liveness as long as `resources/list` lists the URL?
  * `"client"` - URL is for a resource that is managed by the client.
    * The idea here is to enable the model to refer to resources that are managed entirely by the client.  Possibly require a special URL scheme like `mcp-client-storage://{uuid}`?
* A better name for `schemaRef` might be `conformsTo` in order to convey the loose semantics of that property.  Another thought is that it could be a descriptive object instead of a string, e.g. `conformsTo: { jsonSchema: "url" } }`.  (I don't think the value should be the schema object itself, but I suppose forbidding that would be untenable.)

**Reply by evalstate** [reply]
I think these are all good ideas; and with A2A etc. starting to form/be implemented I think it will become easier to make good judgements over what fits best. @tadasant -- I think this would make quite a good CWG topic - shall I set one up?

**Comment by localden** [comment]
We've limited Discussions in this repository to meeting notes and are closing this thread. To continue the conversation, pick a channel from [Contributor Communication](https://modelcontextprotocol.io/community/communication). Thanks for posting.
